modules/DOMStateStorage.js

```diff
@@ -8,18 +8,33 @@ function createKey(key) {
   return KeyPrefix + key
 }
 
+function isHistoryKey(key) {
+  return key.indexOf(KeyPrefix) === 0
+}
+
 export function saveState(key, state) {
   try {
     window.sessionStorage.setItem(createKey(key), JSON.stringify(state))
   } catch (error) {
-    if (error.name === QuotaExceededError || window.sessionStorage.length === 0) {
-      // Probably in Safari "private mode" where sessionStorage quota is 0. #42
-      warning(
-        false,
-        '[history] Unable to save state; sessionStorage is not available in Safari private mode'
-      )
-
-      return
+    if (error.name === QuotaExceededError) {
+      if (window.sessionStorage.length === 0) {
+        // Probably in Safari "private mode" where sessionStorage quota is 0. #42
+        warning(
+          false,
+          '[history] Unable to save state; sessionStorage is not available in Safari private mode'
+        )
+
+        return
+      }
+
+      // Reached quota, remove first (oldest) item in storage
+      for (const storedKey in window.sessionStorage) {
```

Comment:
This loop does not necessarily start with the oldest entry first. AFAICT (just by testing manually in various browser consoles) the iteration order here is not defined. Chrome appears to iterate keys in alphabetical order, while Safari uses a random order.

Since we can't rely on browsers for consistent behavior here, one idea I had was that we could keep our own array in memory, ordered however we want (probably by least recently used) and then just try to persist that array to `sessionStorage` each time. This may be more overhead because we would be serializing a bunch of the same data every time, but it would only be as slow as `JSON.stringify` is on a potentially large array (probably not noticeable).

Another option would be to somehow specify a timestamp in the key itself, so that when we need to remove one we just grab all the keys and sort them to tell which one is the oldest.